### PR TITLE
Add developer documentation for benchmarking

### PR DESCRIPTION
**Pull request by vyasr**
This PR documents best practices for writing cuDF Python benchmarks. It includes an overview of the various fixtures provided by our benchmarking suite to all benchmarks and indicates how best to make use of them. It also discusses the various features of our benchmarking suite (including easy comparison to pandas and running in CI) and what developers must do to maintain compatibility with those features.

A PR to incorporate the [cudf_benchmarks](https://github.com/vyasr/cudf_benchmarks) repo into cudf proper is imminent, but this documentation PR can be reviewed (and merged) independently.

### CODE REVIEW

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+These fixtures should support most use cases.
+Developers may define custom fixtures if necessary, but this should be done with care.
+The default fixtures provide reasonable benchmark coverage without excessive resource usage.
+More bespoke fixtures, if necessary, should be constructed with the same constraints in mind.
+Furthermore, the default fixtures are designed to work when
+[comparing to pandas](pandascompare) or [running tests](testing).
+New fixtures must also account for these use cases.
+
+### The `accepts_cudf_fixture` decorator
```
[the remark this answers is not in the document]
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:85]: Hmm I'm not sure that we want to abstract away the fact that a fixture is being used though. We're trying to hide the mapping between fixtures and parameters, but we expect users to be familiar with pytest and write their tests accordingly, right? cf. [this conversation on the benchmarking thread](https://github.com/rapidsai/cudf/pull/11125#discussion_r903088075), it seems relevant to how we want devs to "think" about these benchmarks.

[the remark this answers is not in the document]
Comment (isVoid) [reply docs/cudf/source/developer_guide/benchmarking.md:85]: Do developers have to learn fixtures in order to make use of this decorator? From the current design, it seems like it has abstracted enough detail so that it's less important whether you mention fixture in the name of the decorator or not. New contributor would look at the decorator:
```python
@accept_cudf_fixture(cls="column, ...)
def bench_col(benchmark, column):
...
```
and learn that some parameterized column will be used to parameterize the test. And that suffices his/her needs. They are certainly welcome to read in depth about its design but I don't think that's not most contributors would do.

[the remark this answers is not in the document]
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:85]: It sounds like both you and @shwina are in favor of a rename, which I support. What name do you think makes sense? `with_frame` is too specific since the object could also be a column. `with_object` seems too generic. What else would you suggest?

[the remark this answers is not in the document]
Comment (shwina) [reply docs/cudf/source/developer_guide/benchmarking.md:85]: - `benchmark_target`?
- `with_cudf_object`?

[the remark this answers is not in the document]
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:85]: After discussing offline, we settled on `benchmark_with_object`.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+Therefore, benchmarks of internal methods of _public_ classes go in files suffixed with `_internal`.
+Benchmarks of `DataFrame._apply_boolean_mask`, for instance, belong in `internal/bench_dataframe_internal.py`.
+```
+
+Free functions have more flexibility.
+Broadly speaking, they should be grouped into benchmark files containing similar functionality.
+For example, I/O benchmarks can all live in `bench_io.py`.
+For now those groupings are left to the discretion of developers.
+
+## Standard fixtures
```
[the remark this answers is not in the document]
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md]: Just pushed some updates, let me know what you think now.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+1. Methods of classes like `DataFrame` or `Series`.
+2. Free functions operating on the above classes like `cudf.merge`.
+
+The former should be organized into files named `bench_class.py`.
+For example, benchmarks of `DataFrame.eval` belong in `API/bench_dataframe.py`.
+Benchmarks should be written at the highest level of generality possible with respect to the class hierarchy.
+For instance, all classes support the `take` method, so those benchmarks belong in `API/bench_frame_or_index.py`.
+If a method has a slightly different API for different classes, benchmarks should use a minimal common API,
+_unless_ developers expect certain arguments to trigger code paths with very different performance characteristics.
```
Comment (shwina) [inline docs/cudf/source/developer_guide/benchmarking.md:35]: Are there any examples we could include here?

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+```{note}
+`pytest` does not support having two benchmark files with the same name, even if they are in separate directories.
+Therefore, benchmarks of internal methods of _public_ classes go in files suffixed with `_internal`.
+Benchmarks of `DataFrame._apply_boolean_mask`, for instance, belong in `internal/bench_dataframe_internal.py`.
```
Comment (shwina) [inline docs/cudf/source/developer_guide/benchmarking.md:40]: Is there value then in having an `internal/` subdirectory rather than flat files?

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+### Benchmark configuration
+
+Our benchmarks aim to support [comparing to pandas](pandascompare) or [running tests](testing) out of the box.
+In order for the former to work, _all tests must import `cudf` and `cupy` from the `config` module_.
+In other words:
+```
+# This is good
+from ..common.config import cudf
+# This is bad
+import cudf
```
Comment (shwina) [inline docs/cudf/source/developer_guide/benchmarking.md:88]:
```suggestion
from ..common.config import cudf  # this is good
import cudf  # this is bad
```

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+```
+
+## Benchmark contents
+
+### Benchmark configuration
+
+Our benchmarks aim to support [comparing to pandas](pandascompare) or [running tests](testing) out of the box.
+In order for the former to work, _all tests must import `cudf` and `cupy` from the `config` module_.
+In other words:
+```
```
Comment (shwina) [inline docs/cudf/source/developer_guide/benchmarking.md:84]:
```suggestion
```python
```

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+Using lazy case functions ensures that the associated objects are only created on an as-needed basis.
+
+When writing cases, just as in writing custom fixtures, developers should be make use of the config variables.
+Cases should import the `NUM_ROWS` and/or `NUM_COLS` variables from the config module and use them to define data sizes.
+
+The observant reader may recognize that cases seem quite familiar to fixtures.
+While the implementations are in fact quite similar, for our purposes it is important to keep the two distinct.
+From a conceptual standpoint, fixtures are homogeneous and generic while cases are heterogeneous and specific.
+Fixtures should be created sparingly and used broadly.
+Cases can be precisely tailored for a single test.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:187]: I think I don't understand this paragraph. At least, it doesn't give me a good idea when I should be using cases over fixtures (or vice versa).

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+### Understanding `benchmark_with_object`
+
+Under the hood, `benchmark_with_object` is made up of two critical pieces, fixture unions and some decorator magic.
+
+#### Fixture unions
+
+Fixture unions are a feature of `pytest_cases`.
+A fixture union is a fixture that, when used as a test function parameter,
+will trigger the test to run once for each fixture contained in the union.
+Since most cuDF benchmarks can be run with the same relative small set of objects,
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:257]:
```suggestion
Since most cuDF benchmarks can be run with the same relatively small set of objects,
```

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+They are not usually necessary for typical developers or benchmark writers.
+This information is primarily for developers looking to extend the types of objects that can be easily benchmarked.
+
+### Understanding `benchmark_with_object`
+
+Under the hood, `benchmark_with_object` is made up of two critical pieces, fixture unions and some decorator magic.
+
+#### Fixture unions
+
+Fixture unions are a feature of `pytest_cases`.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:254]: Link to pytest-cases docs?

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+If a fixture name does not contain a particular component, it represents a union of all values of that component.
+As an example, consider the fixture `dataframe_dtype_int_rows_100`.
+This fixture is a union of both nullable and non-nullable `DataFrame`s of with different numbers of columns.
+
+#### The `benchmark_with_object` decorator
+
+The long names of the above unions are cumbersome when writing tests.
+Moreover, having this information embedded in the name means that in order to change the parameters used,
+the entire benchmark needs to have the fixture name replaced.
+The `benchmark_with_object` decorator is the solution to this problem.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:273]: Link to api docs?

**Review by wence- — COMMENTED** [review]
Minor queries

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 11]
```diff
@@ -0,0 +1,288 @@
+# Benchmarking cuDF
+
+The goal of the benchmarks in this repository is to measure the performance of various cuDF APIs.
+Benchmarks in cuDF are written using the
+[`pytest-benchmark`](https://pytest-benchmark.readthedocs.io/en/latest/index.html) plugin to the
+[`pytest`](https://docs.pytest.org/en/latest/) Python testing framework.
+Using `pytest-benchmark` provides a seamless experience for developers familiar with `pytest`.
+We include benchmarks of both public APIs and internal functions.
+The former give us a macro view of our performance, especially vis a vis pandas.
```
[the remark this answers is not in the document]
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:9]: Fancy

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 3 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:35]: Good idea. I added a mention of `where`. I didn't try to link since the line numbers will change frequently with the addition of new benchmarks, but the mention should be sufficient.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 4 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:40]: I think it's still helpful to clearly delineate all other cases. For example, all `Column` or `ColumnAccessor` benchmarks are internal. The redundancy is only there for public classes with internal methods that we choose to benchmark. I agree that there's some redundancy, though, so I don't feel terribly strongly here if you would prefer to remove.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+### Writing benchmarks 
+
+Just as benchmarks should be written in terms of the highest level classes in the hierarchy,
+they should also assume as little as possible about the nature of the data.
+For instance, unless there are meaningful functional differences,
+benchmarks should not care about the dtype or nullability of the data.
+Objects that differ in these ways should be interchangeable for most benchmarks.
+The goal of writing benchmarks in this way is to then automatically benchmark objects with different properties.
+We support this use case with the `benchmark_with_object` decorator.
```
[the remark this answers is not in the document]
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:103]: Hmm that's tricky. Documentation for internals like this won't get published anywhere, and any link to a source code line will invariably go stale immediately. Maybe just linking to the file containing this? That seems like a reasonably stable URL that should also make it easy to find?

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 9 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:254]: I didn't originally because it's already linked in the parametrization section, but since someone could definitely read this without reading that I agree that it's beneficial to add here too.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 10 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:273]: Same question as above, I'll leave this thread unresolved until we decide what to do on the other one.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (wence-) [reply docs/cudf/source/developer_guide/benchmarking.md:103]: Is the lack of publishing a deliberate choice? While there's benefit in have a curated user-facing API documentation, I think it is also beneficial if every defined function is in some documentation page (it makes stuff a bit more discoverable).

[the remark this answers is not in the document]
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:103]: I wouldn't say a deliberate choice, but I don't know that anyone has ever considered it. I've never seen things like testing utilities that aren't part of the main package included in any published documentation for a package. I am not opposed, but that is a larger task that is probably out of scope for this PR.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 7 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:187]: This discussion is being continued in much greater detail [here](https://github.com/rapidsai/cudf/pull/11180#discussion_r918780717).

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+In order for the former to work, _all tests must import `cudf` and `cupy` from the `config` module_.
+In other words:
+```python
+from ..common.config import cudf  # This is good
+import cudf  # This is bad
+```
+Testing is usually transparently supported except when users define custom fixtures or cases.
+In those instances, as a general rule developers should avoid hardcoding data sizes in benchmarks.
+Instead, data sizes should be dependent on variables stored in `config.py`.
+This requirement is discussed in more depth below.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:93]:
```suggestion
Benchmarks should support both [comparing to pandas](pandascompare) and [being run as tests](testings) out of the box. To satisfy these requirements, one must take a small amount of care when writing a benchmark.

1. Import `cudf` and `cupy` from the config module (this enables swapping out for `pandas` and `numpy` respectively):
    ```python
        from ..common.config import cudf, cupy # Do this
        import cudf, cupy # Not this
    ```
2. Avoid hard-coding benchmark dataset sizes, and instead use the sizes advertised by `config.py` (this enables running the benchmarks in "test" mode on small datasets, which can be much faster)
```

Reasoning: provide the minimal amount of information for what to do (and a short rationale) up front for those people (like me) who are very bad at reading to the end of the documentation page.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (wence-) [reply docs/cudf/source/developer_guide/benchmarking.md:103]: Let's postpone for a broader discussion and leave as is here.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+1. The [`pytest-profiling`](https://github.com/man-group/pytest-plugins/tree/master/pytest-profiling) plugin.
+2. The [`py-spy`](https://github.com/benfred/py-spy) package.
+
+Using the former is as simple as adding the `--profile` (or `--profile-svg`) arguments to your `pytest` invocation.
+The latter requires instead invoking pytest from py-spy, like so:
+```
+py-spy record -- pytest bench_foo.py
+```
+Depending on exactly what information you need, your mileage may vary with each one.
+Developers should try both and see what works for their workflows.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:238]: Minor, but switching between second and third person across sentences is a little jarring.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+@benchmark_with_object(cls="dataframe", dtype="int", cols=6)
+def bench_foo(benchmark, dataframe):
+    benchmark(dataframe.foo)
+```
+
+is functionally equivalent to
+
+```python
+def bench_foo(benchmark, dataframe_dtype_int_cols_6):
+    benchmark(dataframe.foo)
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:287]:
```suggestion
    benchmark(dataframe_dtype_int_cols_6.foo)
```

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+Since most cuDF benchmarks can be run with the same relatively small set of objects,
+our benchmarks generate the Cartesian product of possible fixtures and then create all possible unions.
+
+This feature is critical to the design of our benchmarks.
+For each of the relevant parameter combinations (size, nullability, etc) we programatically generate a new fixture.
+The resulting fixtures are unambiguously named according to the following scheme:
+`{classname}_dtype_{dtype}[_nulls_{true|false}][[_cols_{num_cols}]_rows_{num_rows}]`.
+If a fixture name does not contain a particular component, it represents a union of all values of that component.
+As an example, consider the fixture `dataframe_dtype_int_rows_100`.
+This fixture is a union of both nullable and non-nullable `DataFrame`s of with different numbers of columns.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:266]:
```suggestion
This fixture is a union of both nullable and non-nullable `DataFrame`s with different numbers of columns.
```

**Review by wence- — COMMENTED** [review]
Some more minor comments, but overall looks good I think.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+@pytest.mark.parametrize(
+    "num", [1, 2, 3]
+)
+def bench_foo(benchmark, num):
+    benchmark(num * 2)
+```
+
+for when the parameters are nontrivial and require complex initialization.
+This is common for benchmarks of functions accepting cuDF objects, such as `cudf.concat`.
+With `pytest_cases`, the different cases are instead placed into separate functions and automatically made available.
```
Comment (isVoid) [inline docs/cudf/source/developer_guide/benchmarking.md:153]: Since this is specifically referring to a package, should we leave a link for further reading?

```suggestion
With [`pytest_cases`](https://smarie.github.io/python-pytest-cases/), the different cases are instead placed into separate functions and automatically made available.
```

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+
+This approach is strongly encouraged within cuDF benchmarks.
+It forces developers to put complex initialization into named, documented functions.
+That becomes especially valuable when benchmarking APIs whose performance can vary drastically based on parameters.
+Additionally, cases, like fixtures, are lazily evaluated.
+Initializing complex objects inside a `pytest.mark.parametrize` can dramatically slow down test collection,
+or even lead to out of memory issues if too many complex cases are collected.
+Using lazy case functions ensures that the associated objects are only created on an as-needed basis.
+
+When writing cases, just as in writing custom fixtures, developers should be make use of the config variables.
```
Comment (isVoid) [inline docs/cudf/source/developer_guide/benchmarking.md:180]:
```suggestion
When writing cases, just as in writing custom fixtures, developers should make use of the config variables.
```

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 13 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:93]: Good point. I've adopted your suggestion with some very minor modifications.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 14 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:238]: Agreed, I normally avoid second person but it crept in here unintentionally. I've reworded the sentences to avoid it.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 7 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:187]: I've adapted the conclusions of that discussion into a clear set of guidelines in the latest version of this document.

**Comment by vyasr** [comment]
@shwina @wence- @isVoid thanks for your patience here. I think that I have addressed all the comments. I left open the few discussions that I thought still required some response. I also adapted what we discussed in https://github.com/rapidsai/cudf/pull/11180#discussion_r918780717 into the discussion on how to handle parametrization in different scenarios.

Note that there is significant overlap in some parts of this document with discussions on testing. However, since I anticipate this PR being merged before #11199 I am fine with getting this done and then consolidating the work in that PR.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 4 above)
Comment (shwina) [reply docs/cudf/source/developer_guide/benchmarking.md:42]: I don't feel that strongly either. Resolving.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+These data sizes can be tweaked for debugging purposes (see [](#testing-benchmarks) below).
+Fixture sizes should be relative to the `NUM_ROWS` and/or `NUM_COLS` variables defined in the config module.
+These rules ensure consistency between these fixtures and those provided by `benchmark_with_object`.
+
+In the second case, fixtures are really functioning as parameters, which we discuss in the next section.
+
+
+### Parametrization: custom fixtures, `pytest-cases` and `pytest.mark.parametrize`
+
+When it comes to parametrizing tests, we have a number of options at our disposal.
```
Comment (shwina) [inline docs/cudf/source/developer_guide/benchmarking.md:150]:
```suggestion
When it comes to parametrizing benchmarks, we have a number of options at our disposal.
```

?

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+One option is fixtures, while a second is using `pytest.mark.parametrize`.
+A third option is provided by the [`pytest_cases`](https://smarie.github.io/python-pytest-cases/) `pytest` plugin.
```
Comment (shwina) [inline docs/cudf/source/developer_guide/benchmarking.md:152]: Did we come to a conclusion about whether we want to advocate for the use of `cases` v/s straight parametrize/fixtures? It's not super clear from the [thread](https://github.com/rapidsai/cudf/pull/11180). cc: @wence-

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+  and at least one of those inputs requires complex parametrizations.
+- Given a set of cases, different tests need to run on different subsets with a nonempty intersection.
+
+
+## Comparing to pandas
+
+An important aspect of benchmarking cuDF is comparing it to pandas.
+We often want to generate quantitative comparisons, so we need to make that as easy as possible.
+Our benchmarks support this by setting the environment variable `CUDF_BENCHMARKS_USE_PANDAS`.
+When this variable is detected, all benchmarks will automatically be run using pandas instead of cuDF.
```
Comment (shwina) [inline docs/cudf/source/developer_guide/benchmarking.md:234]: Will this also run the `internal/` benchmarks with Pandas? If not, maybe we could say that here.

File: docs/cudf/source/developer_guide/benchmarking.md [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+Initializing complex objects inside a `pytest.mark.parametrize` can dramatically slow down test collection,
+or even lead to out of memory issues if too many complex cases are collected.
+Using lazy case functions ensures that the associated objects are only created on an as-needed basis.
+
+When writing cases, just as in writing custom fixtures, developers should make use of the config variables.
+Cases should import the `NUM_ROWS` and/or `NUM_COLS` variables from the config module and use them to define data sizes.
+
+Given the plethora of options for parametrization,
+we codify here some best practices for how each should be employed.
+In general, these approach are applicable to parametrizations of different complexity.
```
Comment (wence-) [inline docs/cudf/source/developer_guide/benchmarking.md:197]:
```suggestion
In general, these approaches are applicable to parametrizations of different complexity.
```

**Review by wence- — APPROVED** [review]
Trivial grammar nit, but otherwise looks good, thanks!

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 20 above)
Comment (shwina) [reply docs/cudf/source/developer_guide/benchmarking.md:152]: My fear is that _unless_ it's cut-and-dried, most developers would rather not deal with complexity.

If I develop a feature and need to write a test or benchmark for it, I'm going to choose the path of least resistance, which could include:

- writing as little code as possible
- touching as few files as possible
- learning as few new tools as possible

Comment (shwina) [reply docs/cudf/source/developer_guide/benchmarking.md:152]: Sorry to pile on this discussion further (and late!).

I think the current set of rules for choosing a method of parametrization is well-written and as concise as it can be. 

But, it's quite mind-bending to think about the intersectionality of parameters for someone wanting to develop a feature and write incremental tests or benchmarks for it. Incorporating fixtures into this workflow is easy and familiar as one identifies reusable components between tests. Incorporating cases seems more difficult and really only possible in retrospect - or am I thinking of this all wrong?

Comment (wence-) [reply docs/cudf/source/developer_guide/benchmarking.md:152]: I am not really sufficiently familiar with cases to say, but I think that is reasonable. Probably you would start out writing fixtures, and then, as they became more baroque, wonder if there were a better way (which I believe cases offer).

Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:152]: I don't think that anything that you've said is wrong or untrue, but I think you're classifying as a difference in kind something that is really only a difference in degree.

I would argue that incorporating fixtures effectively is also really only possible in retrospect. As our test suite clearly demonstrates, most people's starting point (at least historically) is to write an unparametrized test with some hardcoded objects, then abuse `pytest.mark.parametrize` to test as many possibilities as you can. Getting everyone to switch to using fixtures already requires some amount of either 1) foresight into what tests you want to write, or 2) refactoring tests in hindsight. Typically the latter occurs as part of PR review, although over the past year or two I think many developers have become more accustomed to thinking in terms of reusable fixtures and reaching for them first.

Cases are just one more step up the complexity ladder. Instead of just thinking about how you might write a bunch of tests that take the exact same arguments, in order to start off with using cases you need to start thinking about tests that might have only partial overlap in parametrization. Otherwise, you refactor tests in hindsight. The review process will train people in those best practices in the same way that we've improved our fixture usage.

> I think the current set of rules for choosing a method of parametrization is well-written and as concise as it can be.

Just to make sure that I understand: am I interpreting this as saying that you're not yet convinced that we should use cases at all?

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 21 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:234]: No it will not since we shouldn't expect compatibility with pandas except in public APIs. I can add that note.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 19 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:150]: Ha yes sorry about that. Thanks for the fix.

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 20 above)
Comment (shwina) [reply docs/cudf/source/developer_guide/benchmarking.md:152]: I did a poor job communicating what changes are required here.

I'm not unilaterally opposed to the use of `cases`, but I feel I need a better understanding of how much friction, if any, we would be adding to the process of developing benchmarks and - more importantly - tests, if we require them.

I think before moving forward with this recommendation, it'd be great to have a deeper discussion about `cases` offline. Perhaps with the rest of the team so we can get their feedback as well?

Comment (wence-) [reply docs/cudf/source/developer_guide/benchmarking.md:152]: FWIW, bringing my experience from other projects that heavily used parameterized fixtures (but not pytest cases). I think there's a tradeoff between generality (and lack of code repetition) that the more complex approaches bring, and ease of debugging (and extraction) of a particular test when something goes wrong and you need to fix things. Although it is relatively easy to run a single test with `pytest test_file::test_name` after it fails for debugging purposes, I have often found that what I end up doing is pulling that test out into a single file that I can run with none of the pytest machinery in the way. If it then depends on a bunch of fixtures this is painful because you have to find all those and so forth.

**Comment by vyasr** [comment]
In the interest of getting this PR merged sooner rather than later, I'm removing the discussion of cases for now until we can have a discussion and come to some consensus on how they should be used. I'm going to copy the exact relevant text out of the current document into this comment and then remove it from the doc so that we can merge and then revisit.

<details><summary>Discussion of cases</summary>

In the second case, fixtures are really functioning as parameters, which we discuss in the next section.

### Parametrization: custom fixtures, `pytest-cases` and `pytest.mark.parametrize`

When it comes to parametrizing tests, we have a number of options at our disposal.
One option is fixtures, while a second is using `pytest.mark.parametrize`.
A third option is provided by the [`pytest_cases`](https://smarie.github.io/python-pytest-cases/) `pytest` plugin.
Our benchmarks make extensive use of this plugin to handle complex parametrization.
Specifically, it provides some syntactic sugar around

```python
@pytest.mark.parametrize(
    "num", [1, 2, 3]
)
def bench_foo(benchmark, num):
    benchmark(num * 2)
```

for when the parameters are nontrivial and require complex initialization.
This is common for benchmarks of functions accepting cuDF objects, such as `cudf.concat`.
With `pytest_cases`, the different cases are instead placed into separate functions and automatically made available.

```python
# bench_foo_cases.py
def case_1():
    return 1

def case_2():
    return 2

def case_3():
    return 3

# bench_foo.py
@pytest_cases.parametrize_with_cases(num)
def bench_foo(benchmark, num):
    benchmark(num * 2)
```

`pytest-cases` is allows developers to put complex initialization into named, documented functions.
That becomes especially valuable when benchmarking APIs whose performance can vary drastically based on parameters.
Additionally, cases, like fixtures, are lazily evaluated.
Initializing complex objects inside a `pytest.mark.parametrize` can dramatically slow down test collection,
or even lead to out of memory issues if too many complex cases are collected.
Using lazy case functions ensures that the associated objects are only created on an as-needed basis.

When writing cases, just as in writing custom fixtures, developers should make use of the config variables.
Cases should import the `NUM_ROWS` and/or `NUM_COLS` variables from the config module and use them to define data sizes.

Given the plethora of options for parametrization,
we codify here some best practices for how each should be employed.
In general, these approach are applicable to parametrizations of different complexity.
For the purpose of this discussion,
we define a "simple parametrization" as parametrization using a list (possibly nested) of primitive objects.
Examples include a list of integers or a list of list of strings.
This _does not_ include e.g. cuDF or pandas objects.

With that in mind, here are some ground rules for how to parametrize.

Use `pytest.mark.parametrize` when:
- One test must be run on many inputs and those inputs are simple to construct.

Use fixtures when:
- One or more tests must be run on the same set of inputs,
  and all of those inputs can be constructed with simple parametrizations.
  In practice, that means that it is acceptable to use a fixture like this:
  ```python
      @pytest.fixture(params=['a', 'b'])
      def foo(request):
          if request.param == 'a':
              # Some complex initialization
          elif request.param == 'b':
              # Some other complex initialization
  ```
  In other words, the construction of the fixture may be complex,
  as long as the parametrization of that construction is simple.

Use `pytest-cases.parametrize_with_cases` when:
- One or more tests must be run on the same set of inputs,
  and at least one of those inputs requires complex parametrizations.
- Given a set of cases, different tests need to run on different subsets with a nonempty intersection.
</details>

File: docs/cudf/source/developer_guide/benchmarking.md (hunk 20 above)
Comment (vyasr) [reply docs/cudf/source/developer_guide/benchmarking.md:152]: I'll leave this conversation unresolved for now so that it's easy to find when we finally come back to this discussion.

**Review by shwina — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge
